### PR TITLE
ISSUE-3313: Add query log write to server handler

### LINKED ISSUE #3313: Handlers call `Interpreter.start()` and  `Interpreter.finish()`
## Summary

Tasks:
- [x] MySQL handler
- [x] ClickHouse handler
- [x] HTTP handler

### LINKED ISSUE #3354: bendctl can not load ontime dataset
**Summary**

Description for this bug.

**How to reproduce**

branch: 
https://github.com/datafuselabs/databend/pull/3321

1. start the databend-query
```
./target/release/databend-query
```

2. create ontime table using bendctl
```
./target/release/bendctl --databend_dir ./.databend --group local load ~/dataset/ontime.csv --table ontime
```

3. load ontime datas
```
./target/release/bendctl --databend_dir ./.databend --group local load ~/dataset/ontime.csv --table ontime
Error: Unknown("Cannot retrieve query result: reqwest::Error { kind: Decode, source: Error(\"expected value\", line: 1, column: 1) }")
```

Server log:
```
./target/release/databend-query 
[2021-12-10T09:22:19Z INFO  databend_query] Config { config_file: "", query: QueryConfig { tenant_id: "", cluster_id: "", num_cpus: 16, mysql_handler_host: "127.0.0.1", mysql_handler_port: 3307, max_active_sessions: 256, clickhouse_handler_host: "127.0.0.1", clickhouse_handler_port: 9000, http_handler_host: "127.0.0.1", http_handler_port: 8000, flight_api_address: "127.0.0.1:9090", http_api_address: "127.0.0.1:8080", metric_api_address: "127.0.0.1:7070", api_tls_server_cert: "", api_tls_server_key: "", api_tls_server_root_ca_cert: "", rpc_tls_server_cert: "", rpc_tls_server_key: "key for rpc server cert", rpc_tls_query_server_root_ca_cert: "", rpc_tls_query_service_domain_name: "localhost", table_engine_csv_enabled: false, table_engine_parquet_enabled: false, table_engine_memory_enabled: true, table_engine_github_enabled: true, wait_timeout_mills: 5000, max_query_log_size: 10000 }, log: LogConfig { log_level: "INFO", log_dir: "./_logs" }, meta: {meta_address: "", meta_user: "", meta_password: "******"}, storage: StorageConfig { storage_type: "disk", disk: DiskStorageConfig { data_path: "_data", temp_data_path: "" }, s3: {s3.storage.region: "", s3.storage.endpoint_url: "", s3.storage.bucket: "", }, azure_storage_blob: {Azure.storage.container: "", } } }
[2021-12-10T09:22:19Z INFO  databend_query] DatabendQuery v-0.1.0-72abc50-simd(1.59.0-nightly-2021-12-10T08:34:37.695798388+00:00)
2021-12-10T09:22:19.773897Z  INFO databend_query::catalogs::impls::mutable_catalog: use embedded meta
[2021-12-10T09:22:19Z INFO  databend_query] MySQL handler listening on 127.0.0.1:3307, Usage: mysql -h127.0.0.1 -P3307
[2021-12-10T09:22:19Z INFO  databend_query] ClickHouse handler listening on 127.0.0.1:9000, Usage: clickhouse-client --host 127.0.0.1 --port 9000
[2021-12-10T09:22:19Z INFO  databend_query] Http handler listening on 127.0.0.1:8000  examples:
    curl --request POST '127.0.0.1:8000/v1/statement/' --header 'Content-Type: text/plain' --data-raw 'SELECT avg(number) FROM numbers(100000000)'
    curl --request POST '127.0.0.1:8000/v1/query/' --header 'Content-Type: application/json' --data-raw '{"sql": "SELECT avg(number) FROM numbers(100000000)"}'
[2021-12-10T09:22:19Z INFO  databend_query] Metric API server listening on 127.0.0.1:7070
[2021-12-10T09:22:19Z WARN  databend_query::api::http_service] Http API TLS not set
2021-12-10T09:22:19.777986Z  INFO poem::server: listening addr=socket://127.0.0.1:8000
2021-12-10T09:22:19.778154Z  INFO poem::server: server started
2021-12-10T09:22:19.778177Z  INFO poem::server: listening addr=socket://127.0.0.1:7070
2021-12-10T09:22:19.778249Z  INFO poem::server: server started
[2021-12-10T09:22:19Z INFO  databend_query] HTTP API server listening on 127.0.0.1:8080
2021-12-10T09:22:19.778589Z  INFO poem::server: listening addr=socket://127.0.0.1:8080
[2021-12-10T09:22:19Z INFO  databend_query] RPC API server listening on 127.0.0.1:9090
2021-12-10T09:22:19.778621Z  INFO poem::server: server started
[2021-12-10T09:22:19Z INFO  databend_query] Databend query has been registered:"" to metasrv:[""].
[2021-12-10T09:22:19Z INFO  databend_query] Ready for connections.
[2021-12-10T09:22:25Z INFO  databend_query::servers::http::v1::http_query_handlers] receive http query: HttpQueryRequest { session: HttpSessionConf { database: None }, sql: "SHOW TABLES LIKE 'ontime';" } PageParams { wait_time: Some(-1) }
[2021-12-10T09:22:25Z INFO  databend_query::sessions::context] Create DatabendQueryContext
[2021-12-10T09:22:26Z INFO  databend_query::sessions::context] Destroy DatabendQueryContext
[2021-12-10T09:22:57Z INFO  databend_query::servers::http::v1::http_query_handlers] receive http query: HttpQueryRequest { session: HttpSessionConf { database: None }, sql: "CREATE TABLE ontime\n(\n    Year                            UInt16,\n    Quarter                         UInt8,\n    Month                           UInt8,\n    DayofMonth                      UInt8,\n    DayOfWeek                       UInt8,\n    FlightDate                      Date,\n    Reporting_Airline               String,\n    DOT_ID_Reporting_Airline        Int32,\n    IATA_CODE_Reporting_Airline     String,\n    Tail_Number                     String,\n    Flight_Number_Reporting_Airline String,\n    OriginAirportID                 Int32,\n    OriginAirportSeqID              Int32,\n    OriginCityMarketID              Int32,\n    Origin                          String,\n    OriginCityName                  String,\n    OriginState                     String,\n    OriginStateFips                 String,\n    OriginStateName                 String,\n    OriginWac                       Int32,\n    DestAirportID                   Int32,\n    DestAirportSeqID                Int32,\n    DestCityMarketID                Int32,\n    Dest                            String,\n    DestCityName                    String,\n    DestState                       String,\n    DestStateFips                   String,\n    DestStateName                   String,\n    DestWac                         Int32,\n    CRSDepTime                      Int32,\n    DepTime                         Int32,\n    DepDelay                        Int32,\n    DepDelayMinutes                 Int32,\n    DepDel15                        Int32,\n    DepartureDelayGroups            String,\n    DepTimeBlk                      String,\n    TaxiOut                         Int32,\n    WheelsOff                       Int32,\n    WheelsOn                        Int32,\n    TaxiIn                          Int32,\n    CRSArrTime                      Int32,\n    ArrTime                         Int32,\n    ArrDelay                        Int32,\n    ArrDelayMinutes                 Int32,\n    ArrDel15                        Int32,\n    ArrivalDelayGroups              Int32,\n    ArrTimeBlk                      String,\n    Cancelled                       UInt8,\n    CancellationCode                String,\n    Diverted                        UInt8,\n    CRSElapsedTime                  Int32,\n    ActualElapsedTime               Int32,\n    AirTime                         Int32,\n    Flights                         Int32,\n    Distance                        Int32,\n    DistanceGroup                   UInt8,\n    CarrierDelay                    Int32,\n    WeatherDelay                    Int32,\n    NASDelay                        Int32,\n    SecurityDelay                   Int32,\n    LateAircraftDelay               Int32,\n    FirstDepTime                    String,\n    TotalAddGTime                   String,\n    LongestAddGTime                 String,\n    DivAirportLandings              String,\n    DivReachedDest                  String,\n    DivActualElapsedTime            String,\n    DivArrDelay                     String,\n    DivDistance                     String,\n    Div1Airport                     String,\n    Div1AirportID                   Int32,\n    Div1AirportSeqID                Int32,\n    Div1WheelsOn                    String,\n    Div1TotalGTime                  String,\n    Div1LongestGTime                String,\n    Div1WheelsOff                   String,\n    Div1TailNum                     String,\n    Div2Airport                     String,\n    Div2AirportID                   Int32,\n    Div2AirportSeqID                Int32,\n    Div2WheelsOn                    String,\n    Div2TotalGTime                  String,\n    Div2LongestGTime                String,\n    Div2WheelsOff                   String,\n    Div2TailNum                     String,\n    Div3Airport                     String,\n    Div3AirportID                   Int32,\n    Div3AirportSeqID                Int32,\n    Div3WheelsOn                    String,\n    Div3TotalGTime                  String,\n    Div3LongestGTime                String,\n    Div3WheelsOff                   String,\n    Div3TailNum                     String,\n    Div4Airport                     String,\n    Div4AirportID                   Int32,\n    Div4AirportSeqID                Int32,\n    Div4WheelsOn                    String,\n    Div4TotalGTime                  String,\n    Div4LongestGTime                String,\n    Div4WheelsOff                   String,\n    Div4TailNum                     String,\n    Div5Airport                     String,\n    Div5AirportID                   Int32,\n    Div5AirportSeqID                Int32,\n    Div5WheelsOn                    String,\n    Div5TotalGTime                  String,\n    Div5LongestGTime                String,\n    Div5WheelsOff                   String,\n    Div5TailNum                     String\n) ENGINE = FUSE;" } PageParams { wait_time: Some(-1) }
[2021-12-10T09:22:57Z INFO  databend_query::sessions::context] Create DatabendQueryContext
2021-12-10T09:22:57.718558Z  INFO common_meta_raft_store::state_machine::sm_meta_api_impl: create table: default: "ontime"
[2021-12-10T09:22:57Z INFO  databend_query::sessions::context] Destroy DatabendQueryContext
[2021-12-10T09:23:00Z INFO  databend_query::servers::http::v1::http_query_handlers] receive http query: HttpQueryRequest { session: HttpSessionConf { database: None }, sql: "SHOW TABLES LIKE 'ontime';" } PageParams { wait_time: Some(-1) }
[2021-12-10T09:23:00Z INFO  databend_query::sessions::context] Create DatabendQueryContext
[2021-12-10T09:23:00Z INFO  databend_query::sessions::context] Destroy DatabendQueryContext

```

### PR DESCRIPTION
**Pull request by BohuTANG**
I hereby agree to the terms of the CLA available at: https://databend.rs/policies/cla/

## Summary

Summary about this PR:
1. Fix DataTime error in quey event
2. Add query log writer to MySQL/ClickHouse/HTTP Handler

## Changelog

- New Feature

## Related Issues

Fixes #3313
Fixes #3354

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

**Review by sundy-li — APPROVED** [review]

File: query/src/interpreters/interpreter_query_log.rs [hunk 1]
```diff
@@ -99,7 +100,7 @@ impl InterpreterQueryLog {
 
         let block = DataBlock::create_by_array(schema.clone(), vec![
             // Type.
-            Series::new(vec![event.log_type as u8]),
+            Series::new(vec![event.log_type as i8]),
```
Comment (sundy-li) [inline query/src/interpreters/interpreter_query_log.rs:103]: So we should support enum type~

File: query/src/servers/http/v1/query/execute_state.rs [hunk 2]
```diff
@@ -196,6 +199,9 @@ impl ExecuteState {
                 }
                 log::debug!("drop block sender!");
             })?;
+
+        // Write Finish to query log table.
+        interpreter.finish().await?;
```
Comment (youngsofun) [inline query/src/servers/http/v1/query/execute_state.rs:204]: @BohuTANG  query is not finished here 
https://github.com/BohuTANG/datafuse/pull/21

Comment (BohuTANG) [reply query/src/servers/http/v1/query/execute_state.rs:204]: Thank you, some unit tests failure related to it , I am trying to fix

File: query/src/servers/http/v1/query/execute_state.rs [hunk 3]
```diff
@@ -107,6 +107,11 @@ impl Executor {
             if kill {
                 r.session.force_kill_query();
             }
+            // Write Finish to query log table.
+            r.interpreter
+                .finish()
+                .await
+                .map_err(|e| log::warn!("interpreter.finish error: {:?}", e));
```
Comment (youngsofun) [inline query/src/servers/http/v1/query/execute_state.rs:114]: sorry, map_err should be unwrap_or_else,  my local test did not finish

Comment (BohuTANG) [reply query/src/servers/http/v1/query/execute_state.rs:115]: Fixed

**Review by zhang2014 — APPROVED** [review]

**Comment by BohuTANG** [comment]
@youngsofun 

This test seems affected, always failed:

```
---- servers::http::http_query_handlers::test_async stdout ----
thread 'servers::http::http_query_handlers::test_async' panicked at 'assertion failed: `(left == right)`

Diff < left / right > :
<None
>Some(
>    "/v1/query/76c8164f-fe15-4738-8d62-52f660c2e9ed/page/0",
>)
```

File: query/tests/it/servers/http/http_query_handlers.rs [hunk 4]
```diff
@@ -80,7 +80,7 @@ async fn test_bad_sql() -> Result<()> {
 async fn test_async() -> Result<()> {
     let sessions = SessionManagerBuilder::create().build()?;
     let route = Route::new().nest("/v1/query", query_route()).data(sessions);
-    let sql = "select sleep(0.1)";
+    let sql = "select sleep(2)";
```
Comment (youngsofun) [inline query/tests/it/servers/http/http_query_handlers.rs:83]: this chage is not needed

File: query/src/servers/http/v1/query/execute_state.rs [hunk 5]
```diff
@@ -157,6 +157,13 @@ impl ExecuteState {
         };
         context.attach_query_str(sql);
 
+        // Auth.
```
Comment (youngsofun) [inline query/src/servers/http/v1/query/execute_state.rs:160]: @BohuTANG you can optionally rebase on [it](https://github.com/datafuselabs/databend/pull/3350) if it is ok to merge fast and first.

**Review by youngsofun — APPROVED** [review]

**Pull request merged**